llvm/lib/ObjectYAML/DXContainerEmitter.cpp

```diff
@@ -263,47 +263,55 @@ void DXContainerWriter::writeParts(raw_ostream &OS) {
     case dxbc::PartType::Unknown:
       break; // Skip any handling for unrecognized parts.
     case dxbc::PartType::RTS0:
-      if (!P.RootSignature.has_value())
-        continue;
-
-      mcdxbc::RootSignatureDesc RS;
-      RS.Flags = P.RootSignature->getEncodedFlags();
-      RS.Version = P.RootSignature->Version;
-      RS.RootParameterOffset = P.RootSignature->RootParametersOffset;
-      RS.NumStaticSamplers = P.RootSignature->NumStaticSamplers;
-      RS.StaticSamplersOffset = P.RootSignature->StaticSamplersOffset;
-
-      for (const auto &Param : P.RootSignature->Parameters) {
-        dxbc::RootParameterHeader Header{Param.Type, Param.Visibility,
-                                         Param.Offset};
-
-        switch (Param.Type) {
-        case llvm::to_underlying(dxbc::RootParameterType::Constants32Bit):
-          dxbc::RootConstants Constants;
-          Constants.Num32BitValues = Param.Constants.Num32BitValues;
-          Constants.RegisterSpace = Param.Constants.RegisterSpace;
-          Constants.ShaderRegister = Param.Constants.ShaderRegister;
-          RS.ParametersContainer.addParameter(Header, Constants);
-          break;
-        case llvm::to_underlying(dxbc::RootParameterType::SRV):
-        case llvm::to_underlying(dxbc::RootParameterType::UAV):
-        case llvm::to_underlying(dxbc::RootParameterType::CBV):
-          dxbc::RTS0::v2::RootDescriptor Descriptor;
-          Descriptor.RegisterSpace = Param.Descriptor.RegisterSpace;
-          Descriptor.ShaderRegister = Param.Descriptor.ShaderRegister;
-          if (RS.Version > 1)
-            Descriptor.Flags = Param.Descriptor.getEncodedFlags();
-          RS.ParametersContainer.addParameter(Header, Descriptor);
-          break;
-        default:
-          // Handling invalid parameter type edge case. We intentionally let
-          // obj2yaml/yaml2obj parse and emit invalid dxcontainer data, in order
-          // for that to be used as a testing tool more effectively.
-          RS.ParametersContainer.addInvalidParameter(Header);
-        }
-      }
-
-      RS.write(OS);
+            if (!P.RootSignature.has_value())
+              continue;
+
+            mcdxbc::RootSignatureDesc RS;
+            RS.Flags = P.RootSignature->getEncodedFlags();
+            RS.Version = P.RootSignature->Version;
+            RS.RootParameterOffset = P.RootSignature->RootParametersOffset;
+            RS.NumStaticSamplers = P.RootSignature->NumStaticSamplers;
+            RS.StaticSamplersOffset = P.RootSignature->StaticSamplersOffset;
+
+            for (DXContainerYAML::RootParameterLocationYaml &L :
+                P.RootSignature->Parameters.Locations) {
+              auto Header = dxbc::RootParameterHeader{
+                  L.Header.Type, L.Header.Visibility, L.Header.Offset};
```

Comment:
I think this would be clearer.
```suggestion
              dxbc::RootParameterHeader Header{
                  L.Header.Type, L.Header.Visibility, L.Header.Offset};
```